solidity/20241126-upgrade-path-testing.md

```diff
@@ -0,0 +1,121 @@
+created by inheriting from `CommonTest`.
+
+# Summary
+
+<!-- Most (if not all) documents should have a summary.
+While the length will likely be proportional to the length of the full document,
+the summary should be as succinct as possible. -->
+
+# Problem Statement + Context
+
+Our test suite is currently not able to replicate the system from a previous release, which
+prevents us from testing the upgrade path to the system currently under development.
+
+As we're moving towards onchain upgrades via the OPCM, we want to be able to test that:
+
+1. The upgrade works to move from the previous release to the system on `develop`.
+2. The upgraded system passes the same set of unit test as the freshly deployed system on `develop`.
+
+# Proposed Solution
+
+## Deploying Superchain contracts with `op-deployer`
+
+We propose to extend `op-deployer bootstrap` to enable deploying superchain contracts.
+
+```shell
+op-deployer bootstrap superchain <artifacts-locator> --outdir <outdir>
+```
+
+This command will write the artifacts to the outdir.
+
+## Setting up the system to test
+
+The current foundry testing uses the following high level call flow.
+
+```mermaid
+graph LR
+    A["CommonTest.setUp()"] --> B["Setup.L1()"] --> C["Deploy.run()"]
+    C --> D["DeployImplementations.run()"]
+    C --> E["DeploySuperchain.run()"]
+    C --> F["OPCM.deploy()"]
+```
+
+This is roughly what each component does:
+
+- **`CommonTest.setUp()`:**
+  - Defines system config settings (ie. `useInterop`).
+  - provides reusable addresses (`alice` and `bob`)
+  - provides reusable generic contracts (`ERC20`)
+- **`Setup.L1()`:**
+  - Reads addresses from the deployments file stored on disk
+  - labels addresses to make traces more readable
+  - ie.
+    ```solidity
+      optimismPortal = IOptimismPortal(deploy.mustGetAddress("OptimismPortalProxy"));
+      vm.label(address(optimismPortal), "OptimismPortal");
+    ```
+- **`Deploy.run()`:**
+  - Deploys all necessary contracts via calls to `DeploySuperchain`, `DeployImplementations`.
+- **`DeploySuperchain.run()`:**
+  - Deploys superchain contracts (`SuperchainConfig` and `ProtocolVersions`).
+- **`DeployImplementations.run()`:**
+  - Deploys contracts (implementations and singletons) necessary for upgrading to the system on
+    `develop`.
+- **`OPCM.deploy()`:**
+  - Deploys all proxies and bespoke singleton contracts as necessary for a new OP Chain.
+
+This work would replace the `Deploy` script with a new `Upgrade` script, resulting in the following
+call flow:
+
+```mermaid
+graph LR
+    A["CommonTest.setUp()"] --> B["Setup.L1()"] -->  C["Upgrade.run()"]
+    C --> D["DeployImplementations.run()"]
+    C --> E["OPCM.upgrade()"]
+```
+
+A description of what new components would do is:
+
+- **`Upgrade.run()`:**
+  - Calls `op-deployer bootstrap superchain` to deploy new superchain contracts (`SuperchainConfig`
+    and `ProtocolVersions`), corresponding to the previous release.
```

Comment:
> corresponding to the previous release.

@mslipper This might be a gotcha I've miseed. ie. since `op-deployer` now only deploys a single release version, we'd need to use the previous release of `op-deployer`, not the one that exists on `develop`. Can we easily install and run a previous version?

Reply:
As long as a tagged release exists that can deploy all the prerequisite contracts, then yes.